src/core/public/plugins/plugin_loader.ts

```diff
@@ -0,0 +1,123 @@
+import { PluginInitializer } from './plugin';
+
+/**
+ * Unknown variant for internal use only for when plugins are not known.
+ * @internal
+ */
+export type UnknownPluginInitializer = PluginInitializer<unknown, Record<string, unknown>>;
+
+/**
+ * Extend window with types for loading bundles
+ * @internal
+ */
+declare global {
+  interface Window {
+    __kbnNonce__: string;
+    __kbnBundles__: {
+      [pluginBundleName: string]: UnknownPluginInitializer | undefined;
+    };
+  }
+}
+
+/**
+ * Timeout for loading a single script in milliseconds.
+ * @internal
+ */
+export const LOAD_TIMEOUT = 120 * 1000; // 2 minutes
+
+/**
+ * Loads the bundle for a plugin onto the page and returns their PluginInitializer. This should
+ * be called for all plugins (once per plugin) in parallel using Promise.all.
+ *
+ * If this is slowing down browser load time, there are some ways we could make this faster:
+ *    - Add these bundles in the generated bootstrap.js file so they're loaded immediately
+ *    - Concatenate all the bundles files on the backend and serve them in single request.
+ *    - Use HTTP/2 to load these bundles without having to open new connections for each.
+ *
+ * This may not be much of an issue since these should be cached by the browser after the first
+ * page load.
+ *
+ * @param basePath
+ * @param plugins
+ * @internal
+ */
+export const loadPluginBundle: LoadPluginBundle = <
+  TSetup,
+  TDependencies extends Record<string, unknown>
+>(
+  addBasePath: (path: string) => string,
+  pluginName: PluginName,
+  { timeoutMs = LOAD_TIMEOUT } = {}
+) =>
+  new Promise<PluginInitializer<TSetup, TDependencies>>((resolve, reject) => {
+    const script = document.createElement('script');
+
+    // Assumes that all plugin bundles get put into the bundles/plugins subdirectory
+    const bundlePath = addBasePath(`/bundles/plugin/${pluginName}.bundle.js`);
+    script.setAttribute('src', bundlePath);
+    script.setAttribute('id', `kbn-plugin-${pluginName}`);
+    script.setAttribute('async', '');
+
+    // Add kbnNonce for CSP
+    script.setAttribute('nonce', window.__kbnNonce__);
+
+    const cleanupTag = () => {
+      clearTimeout(timeout);
+      // Set to null for IE memory leak issue. Webpack does the same thing.
+      // @ts-ignore
+      script.onload = script.onerror = null;
+    };
+
+    // Wire up resolve and reject
+    script.onload = () => {
+      cleanupTag();
+
+      const initializer = window.__kbnBundles__[`plugin/${pluginName}`];
+      if (!initializer || typeof initializer !== 'function') {
+        reject(
+          new Error(`Definition of plugin "${pluginName}" should be a function (${bundlePath}).`)
+        );
+      } else {
+        resolve(initializer as PluginInitializer<TSetup, TDependencies>);
+      }
+    };
+
+    script.onerror = () => {
+      cleanupTag();
+      reject(new Error(`Failed to load "${pluginName}" bundle (${bundlePath})`));
+    };
+
+    const timeout = setTimeout(() => {
```

Comment:
What's the purpose of the custom timeout here versus just relying on errors from the HTTP request?

Reply:
There's two scenarios I can think of where we may not hit Node's default 2 minute timeout for HTTP requests:
- A load balancer may be configured to hold onto requests until a backend is healthy or ready to accept requests.
- A bad connection or server under heavy load may take a long time to transfer the bundle to the client.

In either of those scenarios, I think providing helpful error messages to help troubleshoot why the page is taking so long to load may be useful (Webpack also sets a similar timeout when loading chunks). That said, implementing this does take away some control from the admin for a Kibana instance.